accepted/2024/support-nuget-authentication-plugins-dotnet-tools.md

```diff
@@ -0,0 +1,362 @@
+For example, in the case of the Azure Artifacts Credential Provider, the code looks for `*.exe` or `*.dll` files under the "CredentialProvider.Microsoft" directory.
+However, when .NET tools are installed in the `any` folder, executable files (like `.exe` files on Windows or files with the executable bit set on Linux & Mac) are placed directly in the `any` directory.
+The remaining files are stored in a `.store` folder. This arrangement eliminates the need to search in subdirectories.
+
+NuGet will discover plugins in the new directory by enumerating files, without recursing into child directories.
+This is compatible with the layout that `dotnet tool install` uses, where packages are extracted to a `.store` subdirectory, and shims are put in the tool directory, one file per executable command.
+All files could be considered executables that NuGet will run.
+Alternatively, on Windows, NuGet could filter `*.exe`.
+On Mac and Linux, where apps typically don't have extensions, to do functionally equivalent filtering NuGet should check each file's permissions and ensure the execute bit is enabled (although technically we'd probably also need to check the file owner and group, to know if we should check the user, group, or other permissions).
+
+The new folder structure for NuGet plugins on the Windows platform is as follows:
+
+```folder
+├───any
+│   │   dotnetsay.exe
+│   │
+│   └───.store│
+├───netcore
+├   ├───AWS.CodeArtifact.NuGetCredentialProvider
+├   │
+│   └───CredentialProvider.Microsoft
+└───netfx
+├   ├───AWS.CodeArtifact.NuGetCredentialProvider
+├   │
+├   └───CredentialProvider.Microsoft
+└
+```
+
+The new folder structure for NuGet plugins on the Linux platforms is as follows:
+
+```log
+{user}:~/.nuget/plugins$ dir
+netcore  any
+
+{user}:~/.nuget/plugins$ cd any/
+{user}:~/.nuget/plugins/any$ dir
+botsay  dotnetsay
+
+{user}:~/.nuget/plugins/any$ ls -la
+total 164
+drwxr-xr-x 3 {user}  4096 Feb 10 08:21 .
+drwxr-xr-x 4 {user}  4096 Feb 10 08:20 ..
+drwxr-xr-x 5 {user}  4096 Feb 10 08:21 .store
+-rwxr-xr-x 1 {user} 75632 Feb 10 08:21 botsay
+-rwxr-xr-x 1 {user} 75632 Feb 10 08:20 dotnetsay
+```
+
+## Roadmap
+
+In terms of the implementation roadmap, I propose the following stages:
+
+1. Initially, plugin authors will publish NuGet plugins as .NET Tools. Consumers will install NuGet plugins using `dotnet tool` sub-commands.
+They will specify the plugin's path based on their operating system. We will make the necessary code changes on the NuGet side to enable running the plugins installed as `.NET tools`.
+
+2. Subsequently, we will introduce a `dotnet nuget plugin` subcommand. This will allow users to `install, update, uninstall, and search` for plugins.
+However, we will still rely on the `dotnet tool` command under the hood, as mentioned above.
+Please note that these command names are subject to change based on feedback to the specification we are going to create in the near future.
+
+## Drawbacks
```

Comment:
I just spent a few minutes to test what happens if I 1. install some dotnet tools to a directory then 2. pretend I'm a naive, but well meaning cred provider author who creates a non-.NET provider copying a similar directory structure to the .NET providers. 3. run `dotnet tool list --tool-path .\TestCredProviders`. 

The outcome is that `dotnet tool list` crashes, unhandled exception showing a stack trace, because a subdirectory it expects to be there is not.

Additionally, I tried to create a tool package for a native .exe, making a guess at the contents for the `DotnetToolSettings.xml`. But trying to install this package via `dotnet tool install` failed.

Therefore, a drawback of the design that uses `dotnet tool install` to install packages is that it actively prevents non-.NET credential providers, or at least makes it very difficult for non-.NET cred providers to be installed without causing problems with our tooling.

Note, I'm not saying we need to support non-.NET providers in any way. But if the design for this feature doesn't need to effectively use `dotnet tool install`, `dotnet tool search`, `dotnet tool update`, then we could implement it in a way that doesn't block non-.NET providers, even if such authors wouldn't benefit from all the tooling we provide to make .NET cred providers more productive to implement.

We could take a page out of `dotnet workload`'s book, and have our own install, list, update commands, that do not crash if non-.NET cred providers exist in the `.nuget/plugsin/any` directory. We can re-use `dotnet tool`'s code to create executable entrypoint files from packages to retain the benefit of having this directory a flat list of executable files, making it trivial for NuGet at runtime to discover and launch the app, regardless of what language it was written in.

Reply:
I have updated the specification based on the feedback received during the meeting. Please review the latest spec.

In summary, NuGet will scan the `PATH` environment variable for plugins whose file name begins with `nuget-plugin-` and has the extension `.exe` on Windows, or has the executable bit set on non-Windows platforms.